x-pack/plugins/lens/public/shared_components/coloring/color_ranges/utils/color_ranges_crud.ts

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import { getDataMinMax, roundValue } from '../../utils';
+import { calculateMaxStep } from './utils';
+
+import type { ColorRange, DataBounds, ColorRangeAccessor } from '../types';
+import type { CustomPaletteParamsConfig } from '../../../../../common';
+
+/**
+ * Add new color range after the last item
+ * @internal
+ */
+export const addColorRange = (
```

Comment:
In general all this file has utilities that mutate the original objects and then only shallow copy the wrapping array.
I'd prefer to have a copy of the object before mutating it.